test/e2e/canary_test.go

```diff
@@ -142,3 +147,84 @@ func buildCanaryCurlPod(name, namespace, image, host string) *corev1.Pod {
 		},
 	}
 }
+
+// TestCanaryRouteRotationAnnotation verifies that the
+// canary controller respects the canary route rotation
+// annotation for the default ingress controller.
+//
+// Note that this test will mutate the default ingress controller
+func TestCanaryRouteRotationAnnotation(t *testing.T) {
+	// Set the CanaryRouteRotation annotation to true
+	// on the default ingress controller.
+	if err := setDefaultIngressControllerRotationAnnotation(t, true); err != nil {
+		t.Fatalf("failed to set canary route rotation annotation: %v", err)
+	}
+
+	// Cleanup default ingress controller by setting the canary rotation
+	// annotation to false (and verifying that we were successful in doing so).
+	defer func() {
+		if err := setDefaultIngressControllerRotationAnnotation(t, false); err != nil {
+			t.Fatalf("failed to set canary route rotation annotation: %v", err)
+		}
+	}()
+
+	// Get canary route.
+	canaryRoute := &routev1.Route{}
+	name := controller.CanaryRouteName()
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), name, canaryRoute); err != nil {
+			t.Logf("failed to get canary route %s: %v", name, err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		t.Fatalf("failed to observe canary route: %v", err)
+	}
+
+	// The canary controller should update the canary route after 5 successful
+	// canary checks. Canary checks happen once a minute.
+	updatedCanaryRoute := &routev1.Route{}
+	err = wait.PollImmediate(5*time.Second, 8*time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), name, updatedCanaryRoute); err != nil {
+			t.Logf("failed to get canary route %s: %v", name, err)
+			return false, nil
+		}
+		// If the canaryRoute and updatedCanaryRoute do not have the same targetPort,
+		// then the canary route rotation annotation is working.
+		if cmp.Equal(canaryRoute.Spec.Port.TargetPort, updatedCanaryRoute.Spec.Port.TargetPort) {
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if err != nil {
+		t.Fatalf("failed to observe canary route rotation: %v", err)
+	}
+}
+
+func setDefaultIngressControllerRotationAnnotation(t *testing.T, val bool) error {
+	t.Helper()
```

Comment:
Is this necessary given we only appear to access `t` for logging?

Reply:
Miciah suggested adding this:
```
You can call t.Helper() at the start of this function to cause any log output that this function generates to have the caller's line number.
```

Reply:
I thought this was only useful when you want to fail or error. No big deal.

Reply:
The log messages will have line numbers, right?  The line numbers of the call sites are more specific than line numbers in the helper.